### Задание 1
Импортируйте библиотеку **Pandas** и дайте ей псевдоним **pd**.

In [1]:
import pandas as pd

Создайте датафрейм **authors** со столбцами **author_id** и **author_name**, в которых соответственно содержатся данные: `[1, 2, 3]` и `['Тургенев', 'Чехов', 'Островский']`.

In [2]:
authors = pd.DataFrame({'author_id': [1, 2, 3],
                        'author_name': ['Тургенев', 'Чехов', 'Островский']})
authors

author_id author_name
0          1    Тургенев
1          2       Чехов
2          3  Островский

Затем создайте датафрейм **books** cо столбцами **author_id**, **book_title** и **price**, в которых соответственно содержатся данные: 
- `[1, 1, 1, 2, 2, 3, 3]`, 
- `['Отцы и дети', 'Рудин', 'Дворянское гнездо', 'Толстый и тонкий', 'Дама с собачкой', 'Гроза', 'Таланты и поклонники']`,
- `[450, 300, 350, 500, 450, 370, 290]`.

In [3]:
books = pd.DataFrame({'author_id': [1, 1, 1, 2, 2, 3, 3],
                     'book_title': ['Отцы и дети', 'Рудин', 'Дворянское гнездо', 'Толстый и тонкий', 'Дама с собачкой', 'Гроза', 'Таланты и поклонники'],
                     'price': [450, 300, 350, 500, 450, 370, 290]})
books

author_id            book_title  price
0          1           Отцы и дети    450
1          1                 Рудин    300
2          1     Дворянское гнездо    350
3          2      Толстый и тонкий    500
4          2       Дама с собачкой    450
5          3                 Гроза    370
6          3  Таланты и поклонники    290

### Задание 2
Получите датафрейм **authors_price**, соединив датафреймы **authors** и **books** по полю **author_id**.

In [4]:
authors_price = pd.merge(authors, books, on='author_id', how='inner')
authors_price

author_id author_name            book_title  price
0          1    Тургенев           Отцы и дети    450
1          1    Тургенев                 Рудин    300
2          1    Тургенев     Дворянское гнездо    350
3          2       Чехов      Толстый и тонкий    500
4          2       Чехов       Дама с собачкой    450
5          3  Островский                 Гроза    370
6          3  Островский  Таланты и поклонники    290

### Задание 3
Создайте датафрейм **top5**, в котором содержатся строки из **authors_price** с пятью самыми дорогими книгами.

In [5]:
top5 = authors_price.nlargest(5, 'price')
top5

author_id author_name         book_title  price
3          2       Чехов   Толстый и тонкий    500
0          1    Тургенев        Отцы и дети    450
4          2       Чехов    Дама с собачкой    450
5          3  Островский              Гроза    370
2          1    Тургенев  Дворянское гнездо    350

### Задание 4
Создайте датафрейм **authors_stat** на основе информации из **authors_price**.

В датафрейме **authors_stat** должны быть четыре столбца:
**author_name**, **min_price**, **max_price** и **mean_price**,
в которых должны содержаться соответственно имя автора, минимальная, максимальная и средняя цена на книги этого автора.

In [6]:
authors_stat = authors_price.groupby('author_name').price.agg(min_price='min', max_price='max', mean_price='mean')
authors_stat = authors_stat.reset_index()
authors_stat

author_name  min_price  max_price  mean_price
0  Островский        290        370  330.000000
1    Тургенев        300        450  366.666667
2       Чехов        450        500  475.000000

### Задание 5**
Создайте новый столбец в датафрейме **authors_price** под названием **cover**, в нем будут располагаться данные о том, какая обложка у данной книги - твердая или мягкая. В этот столбец поместите данные из следующего списка:
`['твердая', 'мягкая', 'мягкая', 'твердая', 'твердая', 'мягкая', 'мягкая']`.

In [7]:
authors_price['cover'] = ['твердая', 'мягкая', 'мягкая', 'твердая', 'твердая', 'мягкая', 'мягкая']
authors_price

author_id author_name            book_title  price    cover
0          1    Тургенев           Отцы и дети    450  твердая
1          1    Тургенев                 Рудин    300   мягкая
2          1    Тургенев     Дворянское гнездо    350   мягкая
3          2       Чехов      Толстый и тонкий    500  твердая
4          2       Чехов       Дама с собачкой    450  твердая
5          3  Островский                 Гроза    370   мягкая
6          3  Островский  Таланты и поклонники    290   мягкая

Просмотрите документацию по функции **pd.pivot_table** с помощью вопросительного знака.

In [8]:
?pd.pivot_table

Для каждого автора посчитайте суммарную стоимость книг в твердой и мягкой обложке. Используйте для этого функцию **pd.pivot_table**. При этом столбцы должны называться "твердая" и "мягкая", а индексами должны быть фамилии авторов. Пропущенные значения стоимостей заполните нулями, при необходимости загрузите библиотеку **Numpy**.

In [9]:
book_info = pd.pivot_table(authors_price, values='price', index='author_name', columns='cover', aggfunc='sum', fill_value=0)
book_info

cover        мягкая  твердая
author_name                 
Островский      660        0
Тургенев        650      450
Чехов             0      950

Назовите полученный датасет **book_info** и сохраните его в формат **pickle** под названием **"book_info.pkl"**.

In [10]:
book_info.to_pickle('book_info.pkl')

Затем загрузите из этого файла датафрейм и назовите его **book_info2**. 

In [11]:
book_info2 = pd.read_pickle('book_info.pkl')

Удостоверьтесь, что датафреймы **book_info** и **book_info2** идентичны.

In [12]:
book_info2

cover        мягкая  твердая
author_name                 
Островский      660        0
Тургенев        650      450
Чехов             0      950

In [13]:
book_info.equals(book_info2)

True

### Задание 6**
В этом задании будем работать с датасетом, в котором представлены данные по сериалам и фильмам на **Netflix** (_Netflix Movies and TV Shows_).
Ознакомьтесь с описанием данных и скачайте датасет _netflix_titles.csv_ с сайта **Kaggle.com** https://www.kaggle.com/shivamb/netflix-shows <br>
Считайте датасет. Посмотрите, сколько наблюдений и сколько признаков есть в этом наборе данных.

In [14]:
df = pd.read_csv('Netflix.zip')
df.head()

show_id     type  title           director  \
0      s1  TV Show     3%                NaN   
1      s2    Movie   7:19  Jorge Michel Grau   
2      s3    Movie  23:59       Gilbert Chan   
3      s4    Movie      9        Shane Acker   
4      s5    Movie     21     Robert Luketic   

                                                cast        country  \
0  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil   
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating   duration  \
0    August 14, 2020          2020  TV-MA  4 Seasons   
1  December 23, 2016          2016  TV-MA     93 min   
2  December 20, 2018          2011      R     78 min   
3  November 16, 2017          2009  PG-13     80 min   
4    January 1, 2020          2008  PG-13    123 min   

                                           listed_in  \
0  International TV Shows, TV Dramas, TV Sci-Fi &...   
1                       Dramas, International Movies   
2                Horror Movies, International Movies   
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   

                                         description  
0  In a future where the elite inhabit an island ...  
1  After a devastating earthquake hits Mexico Cit...  
2  When an army recruit is found dead, his fellow...  
3  In a postapocalyptic world, rag-doll robots hi...  
4  A brilliant group of students become card-coun...

In [15]:
df.shape

(7787, 12)

То есть в данном датасете **7787** наблюдений и **12** признаков.

Посчитайте, как много сериалов и фильмов представлено в этом датасете.

In [16]:
df['type'].value_counts()

Movie      5377
TV Show    2410
Name: type, dtype: int64

Посчитайте, как много пропущенных значений (**NaN**) есть в признаках **director** и **country**.

In [17]:
df['director'].isna().sum()

2389

In [18]:
df['country'].isna().sum()

507

Посчитайте, сколько фильмов выпустили раньше _2003_ года.

In [19]:
df[df['release_year'] < 2003].shape[0]

564

Посчитайте, какая страна чаще всего выпускает фильмы на **Netflix**.

In [20]:
df['country'].mode()[0]

'United States'

### Задание 7**
В этом датасете есть признак **date_added** (дата добавления сериала/фильма на платформу), на основе этого признака создайте три новых: **year_added** - год добавления, **month_added** - месяц добавления и **day_added** - день добавления.

In [21]:
df['year_added'] = pd.DatetimeIndex(df['date_added']).year
df['month_added'] = pd.DatetimeIndex(df['date_added']).month
df['day_added'] = pd.DatetimeIndex(df['date_added']).day
df.head()

show_id     type  title           director  \
0      s1  TV Show     3%                NaN   
1      s2    Movie   7:19  Jorge Michel Grau   
2      s3    Movie  23:59       Gilbert Chan   
3      s4    Movie      9        Shane Acker   
4      s5    Movie     21     Robert Luketic   

                                                cast        country  \
0  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil   
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating   duration  \
0    August 14, 2020          2020  TV-MA  4 Seasons   
1  December 23, 2016          2016  TV-MA     93 min   
2  December 20, 2018          2011      R     78 min   
3  November 16, 2017          2009  PG-13     80 min   
4    January 1, 2020          2008  PG-13    123 min   

                                           listed_in  \
0  International TV Shows, TV Dramas, TV Sci-Fi &...   
1                       Dramas, International Movies   
2                Horror Movies, International Movies   
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   

                                         description  year_added  month_added  \
0  In a future where the elite inhabit an island ...      2020.0          8.0   
1  After a devastating earthquake hits Mexico Cit...      2016.0         12.0   
2  When an army recruit is found dead, his fellow...      2018.0         12.0   
3  In a postapocalyptic world, rag-doll robots hi...      2017.0         11.0   
4  A brilliant group of students become card-coun...      2020.0          1.0   

   day_added  
0       14.0  
1       23.0  
2       20.0  
3       16.0  
4        1.0

### Задание 8**
Выберите только наблюдения, где страна `'United States'`, `'India'` или `'United Kingdom'`.

In [22]:
df_filt = df[df['country'].isin(['United States', 'India', 'United Kingdom'])]
df_filt.head()

show_id   type title        director  \
3      s4  Movie     9     Shane Acker   
4      s5  Movie    21  Robert Luketic   
7      s8  Movie   187  Kevin Reynolds   
8      s9  Movie   706   Shravan Kumar   
9     s10  Movie  1920    Vikram Bhatt   

                                                cast        country  \
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   
7  Samuel L. Jackson, John Heard, Kelly Rowan, Cl...  United States   
8  Divya Dutta, Atul Kulkarni, Mohan Agashe, Anup...          India   
9  Rajneesh Duggal, Adah Sharma, Indraneil Sengup...          India   

          date_added  release_year rating duration  \
3  November 16, 2017          2009  PG-13   80 min   
4    January 1, 2020          2008  PG-13  123 min   
7   November 1, 2019          1997      R  119 min   
8      April 1, 2019          2019  TV-14  118 min   
9  December 15, 2017          2008  TV-MA  143 min   

                                           listed_in  \
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   
7                                             Dramas   
8                Horror Movies, International Movies   
9     Horror Movies, International Movies, Thrillers   

                                         description  year_added  month_added  \
3  In a postapocalyptic world, rag-doll robots hi...      2017.0         11.0   
4  A brilliant group of students become card-coun...      2020.0          1.0   
7  After one of his high school students attacks ...      2019.0         11.0   
8  When a doctor goes missing, his psychiatrist w...      2019.0          4.0   
9  An architect and his wife move into a castle t...      2017.0         12.0   

   day_added  
3       16.0  
4        1.0  
7        1.0  
8        1.0  
9       15.0

Для каждой страны посчитайте общее количество фильмов и сериалов по месяцам выпуска. Столбцы нового сводного датафрейма должны называться, как месяца в признаке **month_added**, а индексов будет два: страна и тип.

In [23]:
pd.pivot_table(df_filt, values='title', index=['country', 'type'], aggfunc='count', columns=df['month_added'])

month_added             1.0   2.0   3.0   4.0   5.0   6.0   7.0   8.0   9.0   \
country        type                                                            
India          Movie      65    45    83    72    65    40    69    70    55   
               TV Show     2     2    12    14     5     7     7     4     1   
United Kingdom Movie      17    16    20    12    18    14     7    18    16   
               TV Show    13    17    34    14    10     7    16    22    19   
United States  Movie     245   118   156   139   109   112   144   143   128   
               TV Show    61    45    49    47    46    43    55    55    81   

month_added             10.0  11.0  12.0  
country        type                       
India          Movie      97    87   104  
               TV Show     8     3     6  
United Kingdom Movie      22     8    25  
               TV Show    22     9    20  
United States  Movie     172   210   174  
               TV Show    68    69    80